src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1779,18 +1822,89 @@ RelationRestrictionPartitionKeyIndex(RelationRestriction *relationRestriction)
 
 		if (!targetEntry->resjunk &&
 			IsA(targetExpression, Var) &&
-			IsPartitionColumn(targetExpression, relationPlannerParseQuery))
+			IsPartitionColumn(targetExpression, originalQueryContainingRTEIdentity))
 		{
 			Var *targetColumn = (Var *) targetExpression;
 
-			if (targetColumn->varno == relationRestriction->index)
+/*
+ * As of pg13, columns point to RELATION_RTE so we don't need to do
+ * anything further. Prior to pg13, columns point to JOIN_RTE so we
+ * find the actual column.
+ */
+			RangeTblEntry *rteContainingPartitionKey = NULL;
+			FindReferencedTableColumn(targetExpression, NIL,
+									  originalQueryContainingRTEIdentity,
+									  &targetColumn,
+									  &rteContainingPartitionKey);
+
+			if (rteContainingPartitionKey->rtekind == RTE_RELATION &&
+				GetRTEIdentity(rteContainingPartitionKey) == targetRTEIndex)
 			{
-				return partitionKeyTargetAttrIndex;
+				*partitionKeyIndex = partitionKeyTargetAttrIndex;
+				return (Var *) copyObject(targetColumn);
 			}
 		}
 	}
 
-	return InvalidAttrNumber;
+	return NULL;
+}
+
+
+/*
+ * FindQueryContainingRTEIdentity finds the query/subquery that has an RTE
+ * with rteIndex in its rtable.
+ */
+static Query *
+FindQueryContainingRTEIdentity(Query *mainQuery, int rteIndex)
```

Comment:
minor: `mainQuery` is a little unusual naming for the query. Maybe `query` is enough?